modules/ocf_www/files/build-vhosts

```diff
@@ -14,190 +11,350 @@ from pathlib import Path
+    def dev_alias(self):
+        global dev_config
+        return '{}.{}.ocf.berkeley.edu'.format(
+            self.fqdn.replace('.', '-'),
+            'dev-apphost' if dev_config else 'apphost',
+        )
+
 
 def report(*args, **kwargs):
     print(*args, **kwargs)
     kwargs.setdefault('file', sys.stdout)
     kwargs['file'].flush()
 
 
-def build_config_file():
+def build_config(src_vhosts, template):
     vhosts = set()
 
-    for domain, vhost in get_vhosts().items():
-        ssl_options = {
-            'ssl_key': '/etc/ssl/lets-encrypt/le-vhost.key',
-            'ssl_cert': str(LETS_ENCRYPT_SSL / (domain + '.crt')),
-            'ssl_chain': '/etc/ssl/certs/lets-encrypt.crt',
-        }
-        if (
-                'nossl' in vhost['flags'] or
-                not all(os.path.exists(p) for p in ssl_options.values())
-        ):
-            ssl_options = {}
+    for domain, vhost in src_vhosts.items():
+        user = vhost['username']
+        ssl = SSL(domain)
+        use_ssl = 'nossl' not in vhost['flags'] and ssl.is_valid
 
         # primary vhost
-        user = vhost['username']
         primary = VirtualHost(
             fqdn=domain,
             user=user,
-            relative_docroot=vhost['docroot'],
+            comment='{} (user {})'.format(domain, user),
+            ssl=ssl if use_ssl else None,
+            # only one of {socket,docroot} will ever be present, so this
+            # handles both web and app hosts
+            socket_name=vhost.get('socket', None),
+            relative_docroot=vhost.get('docroot', None),
             redirect_dest=None,
-            ssl_cert=ssl_options.get('ssl_cert'),
-            ssl_key=ssl_options.get('ssl_key'),
-            ssl_chain=ssl_options.get('ssl_chain'),
-            use_hsts='hsts' in vhost['flags'],
-            disabled=user_is_sorried(user),
         )
         vhosts.add(primary)
 
+        # for app vhosts, define a dev vhost as well (HTTP only)
+        if primary.socket_name:
+            vhosts.add(VirtualHost(
+                fqdn=primary.dev_alias,
+                user=user,
+                comment='{} (dev alias of {})'.format(primary.dev_alias, domain),
+                ssl=None,
+                socket_name=primary.socket_name,
+                relative_docroot=None,
+                redirect_dest=None,
+            ))
+
         # if it's an HTTPS domain, add an HTTP domain redirecting to it;
-        # if it has aliases defined, add HTTP domains redirecting there, too
-        for redirect_domain in (
-            ({domain} if ssl_options else set()) | set(vhost['aliases'])
-        ):
+        # if it has aliases defined, add HTTP and HTTPS domains redirects too
+        redirects = set()
+        if use_ssl:
+            redirects.add((primary.fqdn, None))
+        for alias in vhost['aliases']:
+            redirects.add((alias, None))
+            ssl = SSL(alias)
+            if ssl.is_valid:
+                redirects.add((alias, ssl))
```

Comment:
Previously we didn't set up redirects for aliases on port 443, i.e. if you visited `http://xkcd.berkeley.edu` it would redirect you to `xkcd.o.b.e` but not if you visited `https://xkcd.berkeley.edu`.